<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Dataset Field Aggregation • quartzbio.edp</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Dataset Field Aggregation">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">quartzbio.edp</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/r_authentication.html">Authentication</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/quartzbio/quartzbio.edp/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Dataset Field Aggregation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/quartzbio/quartzbio.edp/blob/main/vignettes/aggregating_datasets.Rmd" class="external-link"><code>vignettes/aggregating_datasets.Rmd</code></a></small>
      <div class="d-none name"><code>aggregating_datasets.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Both the EDP Python and R client libraries enable users to perform
aggregations to build complex summaries of data. Aggregation queries can
be run on datasets with the help of facets. Facets can be used to
generate aggregated summaries of string (and date) fields as well as
numeric fields, and they automatically work on top of queries and
filters. Facets can also be nested, which provides an incredibly
efficient mechanism to summarize binned or rolled-up data (i.e. data
summarized by term or by date).</p>
</div>
<div class="section level2">
<h2 id="string-and-date-aggregations">String and Date Aggregations<a class="anchor" aria-label="anchor" href="#string-and-date-aggregations"></a>
</h2>
<p>For string fields (i.e. categorical fields) and date fields, users
can utilize facets to find the total number of unique values as well as
a list of the most common values that occur in the dataset. When used
with a filtered dataset, the results will represent the filtered subset
of data.</p>
<p>The following facet types are supported: - terms (default): Returns a
list of the top terms and the number of times they occur (in order of
this value). The default number of terms returned at once is 10. Users
can set a limit up to 1 million (1,000,000) terms returned. - count:
Returns the number of unique values in the field. For very large
datasets, this is an approximate number.</p>
<p>Facets will not work for text fields that are indexed (and tokenized)
for full-text search. Terms facets are also disabled for _id fields.</p>
<div class="section level3">
<h3 id="examples-in-r">Examples in R:<a class="anchor" aria-label="anchor" href="#examples-in-r"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://quartzbio.github.io/quartzbio.edp/">quartzbio.edp</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load ClinVar</span></span>
<span><span class="va">clinvar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Dataset.get_by_full_path.html">Dataset.get_by_full_path</a></span><span class="op">(</span><span class="st">"quartzbio:Public:/ClinVar/5.2.0-20210110/Variants-GRCH37"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Find the most common genes in ClinVar</span></span>
<span><span class="va">facets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Dataset.facets.html">Dataset.facets</a></span><span class="op">(</span><span class="va">clinvar</span><span class="op">$</span><span class="va">id</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"gene"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Convert the facet results to a matrix</span></span>
<span><span class="va">topGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">facets</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Retrieve the number of unique genes in ClinVar</span></span>
<span><span class="va">count</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Dataset.facets.html">Dataset.facets</a></span><span class="op">(</span><span class="va">clinvar</span><span class="op">$</span><span class="va">id</span>, <span class="st">'{"gene": {"facet_type": "count"}}'</span><span class="op">)</span></span>
<span><span class="co"># Convert the facet result to a number</span></span>
<span><span class="va">count</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter ClinVar for only variants that relate to drug response.</span></span>
<span><span class="co"># Which are the most common genes now?</span></span>
<span><span class="va">filters</span> <span class="op">&lt;-</span> <span class="st">'[["clinical_significance", "Drug response"]]'</span></span>
<span><span class="va">facets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Dataset.facets.html">Dataset.facets</a></span><span class="op">(</span><span class="va">clinvar</span><span class="op">$</span><span class="va">id</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"gene"</span><span class="op">)</span>, filters <span class="op">=</span> <span class="va">filters</span><span class="op">)</span></span>
<span><span class="co"># Convert the facet results to a matrix</span></span>
<span><span class="va">topDrugResponseGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">facets</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># How many genes are in this filtered query?</span></span>
<span><span class="va">count</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Dataset.facets.html">Dataset.facets</a></span><span class="op">(</span><span class="va">clinvar</span><span class="op">$</span><span class="va">id</span>, <span class="st">'{"gene": {"facet_type": "count"}}'</span>, filters <span class="op">=</span> <span class="va">filters</span><span class="op">)</span></span>
<span><span class="co"># Convert the facet result to a number</span></span>
<span><span class="va">countDrugResponseGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Now, get the top 100 most common values</span></span>
<span><span class="va">facets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Dataset.facets.html">Dataset.facets</a></span><span class="op">(</span><span class="va">clinvar</span><span class="op">$</span><span class="va">id</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"gene"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"limit"</span> <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Convert the facet results to a matrix</span></span>
<span><span class="va">top1000DrugResponseGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">facets</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="numeric-aggregations">Numeric Aggregations<a class="anchor" aria-label="anchor" href="#numeric-aggregations"></a>
</h2>
<p>There are various aggregation options are available for numerical
fields (such as float/double, integer/long, and date). Instead of
returning “common terms”, numerical facets can calculate summary
statistics, histograms, and percentiles. The following facet types are
supported:</p>
<ul>
<li>stats: Default stats return average, count, maximum, minimum, and
sum. Extended stats also include standard deviation, standard deviation
lower and upper bounds, sum of squares, and variance.</li>
<li>histogram: values are binned according to a provided interval. For
numerical fields, the default interval is 100. For dates, the default
interval is ‘month’. Histogram intervals must be integers, and will
therefore not work for fields with values between 0 and 1 (such as
allele frequencies).</li>
<li>percentiles: calculates estimated percentiles for a field. By
default, returns the following percentiles: 1, 5, 25, 50, 75, 95, 99.
Percentiles are approximated and have an 1-5% error for very large
datasets.</li>
</ul>
<div class="section level3">
<h3 id="examples-in-r-1">Examples in R:<a class="anchor" aria-label="anchor" href="#examples-in-r-1"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://quartzbio.github.io/quartzbio.edp/">quartzbio.edp</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get ClinVar dataset</span></span>
<span><span class="va">clinvar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Dataset.get_by_full_path.html">Dataset.get_by_full_path</a></span><span class="op">(</span><span class="st">"quartzbio:Public:/ClinVar/5.2.0-20210110/Variants-GRCH37"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get extended statistics for a numerical field.</span></span>
<span><span class="va">stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Dataset.facets.html">Dataset.facets</a></span><span class="op">(</span></span>
<span>  <span class="va">clinvar</span><span class="op">$</span><span class="va">id</span>,</span>
<span>  <span class="st">'{"info.ALLELEID": {</span></span>
<span><span class="st">        "facet_type": "stats", "extended": true}}'</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Get the min and max values</span></span>
<span><span class="va">stats</span><span class="op">$</span><span class="va">review_status_star</span><span class="op">$</span><span class="va">min</span></span>
<span><span class="va">stats</span><span class="op">$</span><span class="va">review_status_star</span><span class="op">$</span><span class="va">max</span></span>
<span></span>
<span><span class="co"># Calculate a histogram for genomic position in chromosome 12</span></span>
<span><span class="va">facets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Dataset.facets.html">Dataset.facets</a></span><span class="op">(</span></span>
<span>  <span class="va">clinvar</span><span class="op">$</span><span class="va">id</span>,</span>
<span>  <span class="st">'{"genomic_coordinates.start": {"facet_type": "histogram"}}'</span>,</span>
<span>  filters <span class="op">=</span> <span class="st">'[["genomic_coordinates.chromosome", 12]]'</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Convert the result to a matrix</span></span>
<span><span class="va">genomicCoordinates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">facets</span><span class="op">$</span><span class="va">genomic_coordinates.start</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="nested-aggregations">Nested Aggregations<a class="anchor" aria-label="anchor" href="#nested-aggregations"></a>
</h2>
<p>Nested aggregations can be used to apply an aggregation query to the
result of another aggregation query. For example, given a dataset with
patient information, users may want to determine the most common
diagnosis age for each cancer type. Users could iterate through each
cancer type and run a facets query on the age field, but that would
require a large number of expensive API calls. Instead, by using nested
aggregations, users can simply construct a facets query within an
existing facets query, as in the example below.</p>
<p>At this time, users may only nest term and histogram facets under
terms facets. Nesting within histogram facets is not currently
supported.</p>
<p>The following example yields the top ten genes associated with each
disease in the public TCGA somatic mutations dataset:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://quartzbio.github.io/quartzbio.edp/">quartzbio.edp</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Retrieve the TCGA Somatic Mutations dataset</span></span>
<span><span class="va">TCGA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Dataset.get_by_full_path.html">Dataset.get_by_full_path</a></span><span class="op">(</span><span class="st">"quartzbio:Public:/TCGA/2.0.0-2018-mc3-v0.2.8/SomaticMutations-GRCh37"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Retrieve each disease (terms facets)</span></span>
<span><span class="co"># and the gene for each (nested terms facet)</span></span>
<span><span class="va">facets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"disease"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"limit"</span> <span class="op">=</span> <span class="fl">100</span>,</span>
<span>    <span class="st">"facets"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      <span class="st">"hugo_symbol"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"limit"</span> <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Dataset.facets.html">Dataset.facets</a></span><span class="op">(</span><span class="va">TCGA</span><span class="op">$</span><span class="va">id</span>, facets <span class="op">=</span> <span class="va">facets</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert results to data frame</span></span>
<span><span class="va">data_frame</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">results</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View data frame</span></span>
<span><span class="va">data_frame</span></span>
<span></span>
<span><span class="co"># Output:</span></span>
<span><span class="co"># The disease is the first value in each row, followed by the total number of records and then each gene and its count</span></span>
<span></span>
<span><span class="co"># disease UCEC, 934029, TTN, 2961, MUC16, 863, PTEN, 684, DST, 677, SYNE1, 642, CSMD3, 628, RYR2, 613, NEB, 569, ZFHX4, 551, OBSCN, 549</span></span>
<span></span>
<span><span class="co"># disease SKCM, 494062, TTN, 3059, MUC16, 2022, DNAH5, 885, PCLO, 672, LRP1B, 515, ANK3, 494, GPR98, 482, CSMD1, 472, DNAH7, 460, CSMD2, 453</span></span>
<span></span>
<span><span class="co"># disease COAD, 240187, TTN, 882, APC, 461, MUC16, 323, SYNE1, 300, TP53, 229, OBSCN, 220, FAT4, 211, RYR2, 186, NEB, 176, KRAS, 171</span></span>
<span></span>
<span><span class="co"># disease LUAD, 222076, TTN, 935, MUC16, 562, RYR2, 488, CSMD3, 480, LRP1B, 392, USH2A, 374, ZFHX4, 348, TP53, 304, FLG, 269, XIRP2, 268</span></span>
<span></span>
<span><span class="co"># disease STAD, 217398, TTN, 882, MUC16, 343, SYNE1, 227, LRP1B, 218, TP53, 216, FAT4, 201, OBSCN, 200, FLG, 193, CSMD3, 192, PCLO, 169</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Caplan, Karl Forner, QuartzBio.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Last modified: 26-Mar-2025 Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
