<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Expression Functions • quartzbio.edp</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Expression Functions">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">quartzbio.edp</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/r_authentication.html">Authentication</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/quartzbio/quartzbio.edp/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Expression Functions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/quartzbio/quartzbio.edp/blob/main/vignettes/expression_functions.Rmd" class="external-link"><code>vignettes/expression_functions.Rmd</code></a></small>
      <div class="d-none name"><code>expression_functions.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>EDP expressions can use Python-like functions to pull data from any
dataset, calculate statistics, or run advanced algorithms. Users are
recommended to read the Expressions documentation for an in-depth review
of use cases.</p>
</div>
<div class="section level2">
<h2 id="function-list">Function List<a class="anchor" aria-label="anchor" href="#function-list"></a>
</h2>
<p>All available functions are listed below:</p>
<table class="table">
<colgroup>
<col width="24%">
<col width="24%">
<col width="50%">
</colgroup>
<thead><tr class="header">
<th align="center">Function</th>
<th align="center">Data Type</th>
<th align="center">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">annotate</td>
<td align="center">object</td>
<td align="center">Annotate a record with a template. annotate(record,
template, debug, include_errors)</td>
</tr>
<tr class="even">
<td align="center">beacon</td>
<td align="center">object</td>
<td align="center">Retrieves the beacon results for any entity.
beacon(entity, entity_type, beacon_set, datasets, visibility)</td>
</tr>
<tr class="odd">
<td align="center">classify_variant</td>
<td align="center">object</td>
<td align="center">Classify a variant using one of multiple classifiers.
classify_variant(variant, classifier)</td>
</tr>
<tr class="even">
<td align="center">coerce_list</td>
<td align="center">auto (list)</td>
<td align="center">Coerce a value to a list. Single items will become a
single value list. Lists will remain lists. None will return an empty
list. coerce_list(value)</td>
</tr>
<tr class="odd">
<td align="center">concat</td>
<td align="center">string</td>
<td align="center">Combine text from multiple lists or strings.
concat(values, delimiter)</td>
</tr>
<tr class="even">
<td align="center">crossmap</td>
<td align="center">string</td>
<td align="center">Convert a variant or genomic region entity between
different genome builds using the Ensembl CrossMap tool. The
functionality of this expression is the same as UCSC’s liftOver tool.
crossmap(entity, target_build)</td>
</tr>
<tr class="odd">
<td align="center">dataset_count</td>
<td align="center">integer</td>
<td align="center">Calculate the total number of results (or “hits”) for
a given query. Returns the number of results. dataset_count(dataset,
entities, filters, query)</td>
</tr>
<tr class="even">
<td align="center">dataset_entity_top_terms</td>
<td align="center">string (list)</td>
<td align="center">Retrieve the top entities for any entity field in a
dataset. Returns a list of strings, in order of occurrence or None if
the dataset can not be queried by this entity.
dataset_entity_top_terms(dataset, entity, limit, filters, query)</td>
</tr>
<tr class="odd">
<td align="center">dataset_field_percentiles</td>
<td align="center">object</td>
<td align="center">Calculates the percentiles for any integer field.
Returns an object containing the desired percentiles.
dataset_field_percentiles(dataset, field, percents, entities, filters,
query)</td>
</tr>
<tr class="even">
<td align="center">dataset_field_stats</td>
<td align="center">object</td>
<td align="center">Calculates statistics for any numeric field. Returns
an object containing field statistics. dataset_field_stats(dataset,
field, entities, filters, query)</td>
</tr>
<tr class="odd">
<td align="center">dataset_field_terms_count</td>
<td align="center">integer</td>
<td align="center">Retrieve the number of unique terms for any string
field in a dataset. Returns the number of unique terms.
dataset_field_terms_count(dataset, field, entities, filters, query)</td>
</tr>
<tr class="even">
<td align="center">dataset_field_top_terms</td>
<td align="center">object (list)</td>
<td align="center">Retrieve the top terms for any string field in a
dataset. Returns a list of objects containing the term and number of
times it occurs, in order of occurrence.
dataset_field_top_terms(dataset, field, limit, entities, filters,
query)</td>
</tr>
<tr class="odd">
<td align="center">dataset_field_values</td>
<td align="center">auto (list)</td>
<td align="center">Retrieves a list of non-empty values for a dataset
field. Returns a list of values from the specified field.
dataset_field_values(dataset, field, limit, entities, filters,
query)</td>
</tr>
<tr class="even">
<td align="center">dataset_query</td>
<td align="center">object (list)</td>
<td align="center">Query any dataset with optional filters and/or
entities. Returns a list of results. dataset_query(dataset, fields,
limit, entities, filters, query)</td>
</tr>
<tr class="odd">
<td align="center">datetime_format</td>
<td align="center">string</td>
<td align="center">Format datetime strings. By default, it returns an
ISO 8601 format date time string. To override, provide an optional
input_format or output_format to be used. datetime_format(value,
input_format, output_format)</td>
</tr>
<tr class="even">
<td align="center">entity_ids</td>
<td align="center">string</td>
<td align="center">Retrieve one or more entity IDs for a query.
entity_ids(entity_type, entity)</td>
</tr>
<tr class="odd">
<td align="center">error</td>
<td align="center">error</td>
<td align="center">Raise a FunctionError error(message)</td>
</tr>
<tr class="even">
<td align="center">explode</td>
<td align="center">object (list)</td>
<td align="center">Split N values from M list fields into N records. If
_id is in the original record, each new record will have an integer
appended to the _id with the index of each exploded record.
explode(record, fields)</td>
</tr>
<tr class="odd">
<td align="center">findall</td>
<td align="center">string (list)</td>
<td align="center">Returns all non-overlapping matches of pattern in
string, as a list of strings. The string is scanned left-to-right, and
matches are returned in the order found. If one or more groups are
present in the pattern, returns a list of groups. findall(pattern,
string, regex_ignorecase, regex_dotall, regex_multiline)</td>
</tr>
<tr class="even">
<td align="center">genomic_sequence</td>
<td align="center">string</td>
<td align="center">Retrieves a specific sequence from the genome.
genomic_sequence(genomic_region)</td>
</tr>
<tr class="odd">
<td align="center">get</td>
<td align="center">auto</td>
<td align="center">Get the value at any depth of a nested object based
on the path described by path. If path doesn’t exist, default is
returned. get(obj, path, default)</td>
</tr>
<tr class="even">
<td align="center">melt</td>
<td align="center">object (list)</td>
<td align="center">Convert a wide dataset to a long dataset by “melting”
one or more fields into “key” and “value” fields. All fields must have
the same data type. melt(record, fields, key_field, value_field,
melt_list_values)</td>
</tr>
<tr class="odd">
<td align="center">normalize_aa_change</td>
<td align="center">string</td>
<td align="center">Normalize an amino acid change (beta)
normalize_aa_change(aa_change, ref, alt)</td>
</tr>
<tr class="even">
<td align="center">normalize_variant</td>
<td align="center">string</td>
<td align="center">Normalize a variant ID (minimal representation and
left shifting). normalize_variant(variant)</td>
</tr>
<tr class="odd">
<td align="center">now</td>
<td align="center">string</td>
<td align="center">Retrieves the current date and time. now(timezone,
template)</td>
</tr>
<tr class="even">
<td align="center">predict_variant_effects</td>
<td align="center">object (list)</td>
<td align="center">Predict the effects of a variant using Veppy.
predict_variant_effects(variant, default_transcript, gene_model)</td>
</tr>
<tr class="odd">
<td align="center">prevalence</td>
<td align="center">double</td>
<td align="center">Calculates the frequency that a value occurs within a
population. Typically used to calculate the prevalence of variants or
genes across samples in a dataset. Returns the frequency of occurrence.
Please note: in large datasets the result is approximate and can have an
error of up to 5%. prevalence(dataset, entity, sample_field,
filters)</td>
</tr>
<tr class="even">
<td align="center">search</td>
<td align="center">boolean</td>
<td align="center">Scan through string looking for the first location
where the regular expression pattern produces a match. Returns True on a
match and False if no position in the string matches the pattern.
search(pattern, string, regex_ignorecase, regex_dotall,
regex_multiline)</td>
</tr>
<tr class="odd">
<td align="center">search_groups</td>
<td align="center">string (list)</td>
<td align="center">Scan through string looking for the first location
where the regular expression pattern produces a match. Returns a list of
strings corresponding to the groups in the pattern.
search_groups(pattern, string, regex_ignorecase, regex_dotall,
regex_multiline)</td>
</tr>
<tr class="even">
<td align="center">split</td>
<td align="center">string (list)</td>
<td align="center">Split text based on a delimiter and optionally strip
whitespace. split(value, delimiter, regex, strip, regex_ignorecase,
regex_dotall, regex_multiline)</td>
</tr>
<tr class="odd">
<td align="center">sub</td>
<td align="center">string</td>
<td align="center">Return the string obtained by replacing the leftmost
non-overlapping occurrences of pattern in string by the replacement
repl. If the pattern isn’t found, string is returned unchanged.
sub(pattern, repl, string, count, regex_ignorecase, regex_dotall,
regex_multiline)</td>
</tr>
<tr class="even">
<td align="center">tabulate</td>
<td align="center">object (list)</td>
<td align="center">Converts a list of objects into a table (i.e. a
two-dimensional array). tabulate(objects, fields, header)</td>
</tr>
<tr class="odd">
<td align="center">today</td>
<td align="center">string</td>
<td align="center">Returns the current date. today(timezone,
template)</td>
</tr>
<tr class="even">
<td align="center">translate_variant</td>
<td align="center">object</td>
<td align="center">Translate variant into a protein change.
translate_variant(variant, gene_model, transcript, include_effects)</td>
</tr>
<tr class="odd">
<td align="center">user</td>
<td align="center">object</td>
<td align="center">Returns the currently authenticated user. user()</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="function-details">Function Details<a class="anchor" aria-label="anchor" href="#function-details"></a>
</h2>
<div class="section level3">
<h3 id="annotate">annotate<a class="anchor" aria-label="anchor" href="#annotate"></a>
</h3>
<p>Annotate a record with a template. Output data type: object</p>
<div class="section level5">
<h5 id="syntax">Syntax<a class="anchor" aria-label="anchor" href="#syntax"></a>
</h5>
<pre><code><span><span class="fu">annotate</span><span class="op">(</span><span class="va">record</span>, <span class="va">template</span>, <span class="va">debug</span>, <span class="va">include_errors</span><span class="op">)</span></span></code></pre>
<ul>
<li>record: (object) The record to be annotated</li>
<li>template: (str) The ID of the template</li>
<li>debug: (bool) Enable debug mode (default: False)</li>
<li>include_errors: (bool) Include errors in output (default: True)</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="beacon">beacon<a class="anchor" aria-label="anchor" href="#beacon"></a>
</h3>
<p>Retrieves the beacon results for any entity. Output data type:
object. Output object properties:</p>
<ul>
<li>failed_count: The number of datasets that failed (timed-out)</li>
<li>failed: List of datasets that failed (timed-out)</li>
<li>not_found_count: The number of datasets without results</li>
<li>found_count: The number of datasets with results</li>
<li>found: List of datasets with results</li>
<li>not_found: List of datasets without results</li>
</ul>
<div class="section level5">
<h5 id="syntax-1">Syntax<a class="anchor" aria-label="anchor" href="#syntax-1"></a>
</h5>
<pre><code><span><span class="fu">beacon</span><span class="op">(</span><span class="va">entity</span>, <span class="va">entity_type</span>, <span class="va">beacon_set</span>, <span class="va">datasets</span>, <span class="va">visibility</span><span class="op">)</span></span></code></pre>
<ul>
<li>entity: The entity value</li>
<li>entity_type: A valid entity type</li>
<li>beacon_set (optional): A valid beacon set ID</li>
<li>datasets (optional): A list of datasets to beacon</li>
<li>visibility (optional): Which datasets to beacon (default:
vault)</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="classify_variant">classify_variant<a class="anchor" aria-label="anchor" href="#classify_variant"></a>
</h3>
<p>Classify a variant using one of multiple classifiers. Output data
type: object</p>
<div class="section level5">
<h5 id="syntax-2">Syntax<a class="anchor" aria-label="anchor" href="#syntax-2"></a>
</h5>
<pre><code><span><span class="fu">classify_variant</span><span class="op">(</span><span class="va">variant</span>, <span class="va">classifier</span><span class="op">)</span></span></code></pre>
<ul>
<li>variant: The variant</li>
<li>classifier: The desired classifier (default: “germline”)</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="coerce_list">coerce_list<a class="anchor" aria-label="anchor" href="#coerce_list"></a>
</h3>
<p>Coerce a value to a list. Single items will become a single value
list. Lists will remain lists. None will return an empty list. Output
data type: auto (list)</p>
<div class="section level5">
<h5 id="syntax-3">Syntax<a class="anchor" aria-label="anchor" href="#syntax-3"></a>
</h5>
<pre><code><span><span class="fu">coerce_list</span><span class="op">(</span><span class="va">value</span><span class="op">)</span></span></code></pre>
<ul>
<li>value: The value to coerce to a list</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="concat">concat<a class="anchor" aria-label="anchor" href="#concat"></a>
</h3>
<p>Combine text from multiple lists or strings. Output data type:
string</p>
<div class="section level5">
<h5 id="syntax-4">Syntax<a class="anchor" aria-label="anchor" href="#syntax-4"></a>
</h5>
<pre><code><span><span class="fu">concat</span><span class="op">(</span><span class="va">values</span>, <span class="va">delimiter</span><span class="op">)</span></span></code></pre>
<ul>
<li>values: The list of values to concatenate</li>
<li>delimiter (default: ““): The character to use in between values</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="crossmap">crossmap<a class="anchor" aria-label="anchor" href="#crossmap"></a>
</h3>
<p>Convert a variant or genomic region entity between different genome
builds using the Ensembl CrossMap tool. The functionality of this
expression is the same as UCSC’s liftOver tool. Output data type:
string</p>
<div class="section level5">
<h5 id="syntax-5">Syntax<a class="anchor" aria-label="anchor" href="#syntax-5"></a>
</h5>
<pre><code><span><span class="fu">crossmap</span><span class="op">(</span><span class="va">entity</span>, <span class="va">target_build</span><span class="op">)</span></span></code></pre>
<ul>
<li>entity: The entity (either a valid quartzbio variant
BUILD-CHROMOSOME-START-STOP-ALT or genomic region
BUILD-CHROMOSOME-START-STOP)</li>
<li>target_build: The target genome build (GRCH37 or GRCH38)</li>
</ul>
</div>
<div class="section level5">
<h5 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h5>
<p><code>crossmap("GRCH38-13-32338647-32338647-T", "GRCH37")</code></p>
</div>
</div>
<div class="section level3">
<h3 id="dataset_count">dataset_count<a class="anchor" aria-label="anchor" href="#dataset_count"></a>
</h3>
<p>Calculate the total number of results (or “hits”) for a given query.
Returns the number of results. Output data type: integer</p>
<div class="section level5">
<h5 id="syntax-6">Syntax<a class="anchor" aria-label="anchor" href="#syntax-6"></a>
</h5>
<pre><code><span><span class="fu">dataset_count</span><span class="op">(</span><span class="va">dataset</span>, <span class="va">entities</span>, <span class="va">filters</span>, <span class="va">query</span><span class="op">)</span></span></code></pre>
<ul>
<li>dataset: Any dataset with query permissions</li>
<li>entities (optional): A list of entity tuples: [(entity_type,
entity)]</li>
<li>filters (optional): A valid filter block</li>
<li>query (optional): A query string</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="dataset_entity_top_terms">dataset_entity_top_terms<a class="anchor" aria-label="anchor" href="#dataset_entity_top_terms"></a>
</h3>
<p>Retrieve the top entities for any entity field in a dataset. Returns
a list of strings, in order of occurrence or None if the dataset can not
be queried by this entity. Output data type: string (list)</p>
<div class="section level5">
<h5 id="syntax-7">Syntax<a class="anchor" aria-label="anchor" href="#syntax-7"></a>
</h5>
<pre><code><span><span class="fu">dataset_entity_top_terms</span><span class="op">(</span><span class="va">dataset</span>, <span class="va">entity</span>, <span class="va">limit</span>, <span class="va">filters</span>, <span class="va">query</span><span class="op">)</span></span></code></pre>
<ul>
<li>dataset: Any dataset with query permissions</li>
<li>entity: The entity_type to return within the dataset</li>
<li>limit (optional): The number of terms to retrieve (default:
1000)</li>
<li>filters (optional): Dataset filters</li>
<li>query (optional): A query string</li>
</ul>
</div>
<div class="section level5">
<h5 id="examples-1">Examples<a class="anchor" aria-label="anchor" href="#examples-1"></a>
</h5>
<p><code>dataset_entity_top_terms("quartzbio:public:/ClinVar/5.1.0-20200720/Variants-GRCH38", "gene")</code></p>
</div>
</div>
<div class="section level3">
<h3 id="dataset_field_percentiles">dataset_field_percentiles<a class="anchor" aria-label="anchor" href="#dataset_field_percentiles"></a>
</h3>
<p>Calculates the percentiles for any integer field. Returns an object
containing the desired percentiles. Output data type: object</p>
<div class="section level5">
<h5 id="syntax-8">Syntax<a class="anchor" aria-label="anchor" href="#syntax-8"></a>
</h5>
<pre><code><span><span class="fu">dataset_field_percentiles</span><span class="op">(</span><span class="va">dataset</span>, <span class="va">field</span>, <span class="va">percents</span>, <span class="va">entities</span>, <span class="va">filters</span>, <span class="va">query</span><span class="op">)</span></span></code></pre>
<ul>
<li>dataset: Any dataset with query permissions</li>
<li>field: The field within the dataset</li>
<li>percents: The percentiles to calculate (default: 1, 5, 25, 50, 75,
95, 99)</li>
<li>entities (optional): A list of entity tuples: [(entity_type,
entity)]</li>
<li>filters (optional): Dataset filters</li>
<li>query (optional): A query string</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="dataset_field_stats">dataset_field_stats<a class="anchor" aria-label="anchor" href="#dataset_field_stats"></a>
</h3>
<p>Calculates statistics for any numeric field. Returns an object
containing field statistics. Output data type: object. Output object
properties:</p>
<ul>
<li>count: The total number of values</li>
<li>max: The maximum value observed</li>
<li>sum: The sum of all values</li>
<li>avg: The average value</li>
<li>min: The minimum value observed</li>
</ul>
<div class="section level5">
<h5 id="syntax-9">Syntax<a class="anchor" aria-label="anchor" href="#syntax-9"></a>
</h5>
<pre><code><span><span class="fu">dataset_field_stats</span><span class="op">(</span><span class="va">dataset</span>, <span class="va">field</span>, <span class="va">entities</span>, <span class="va">filters</span>, <span class="va">query</span><span class="op">)</span></span></code></pre>
<ul>
<li>dataset: Any dataset with query permissions</li>
<li>field: The field within the dataset</li>
<li>entities (optional): A list of entity tuples: [(entity_type,
entity)]</li>
<li>filters (optional): Dataset filters</li>
<li>query (optional): A query string</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="dataset_field_terms_count">dataset_field_terms_count<a class="anchor" aria-label="anchor" href="#dataset_field_terms_count"></a>
</h3>
<p>Retrieve the number of unique terms for any string field in a
dataset. Returns the number of unique terms. Output data type:
integer</p>
<div class="section level5">
<h5 id="syntax-10">Syntax<a class="anchor" aria-label="anchor" href="#syntax-10"></a>
</h5>
<pre><code><span><span class="fu">dataset_field_terms_count</span><span class="op">(</span><span class="va">dataset</span>, <span class="va">field</span>, <span class="va">entities</span>, <span class="va">filters</span>, <span class="va">query</span><span class="op">)</span></span></code></pre>
<ul>
<li>dataset: Any dataset with query permissions</li>
<li>field: The field within the dataset</li>
<li>entities (optional): A list of entity tuples: [(entity_type,
entity)]</li>
<li>filters (optional): Dataset filters</li>
<li>query (optional): A query string</li>
</ul>
</div>
<div class="section level5">
<h5 id="examples-2">Examples<a class="anchor" aria-label="anchor" href="#examples-2"></a>
</h5>
<p><code>dataset_field_terms_count("quartzbio:public:/ClinVar/5.1.0-20200720/Variants-GRCh38", "clinical_significance")</code></p>
</div>
</div>
<div class="section level3">
<h3 id="dataset_field_top_terms">dataset_field_top_terms<a class="anchor" aria-label="anchor" href="#dataset_field_top_terms"></a>
</h3>
<p>Retrieve the top terms for any string field in a dataset. Returns a
list of objects containing the term and number of times it occurs, in
order of occurrence. Output data type: object (list). Output object
properties:</p>
<ul>
<li>count: Number of times it occurs</li>
<li>term: Term value</li>
</ul>
<div class="section level5">
<h5 id="syntax-11">Syntax<a class="anchor" aria-label="anchor" href="#syntax-11"></a>
</h5>
<pre><code><span><span class="fu">dataset_field_top_terms</span><span class="op">(</span><span class="va">dataset</span>, <span class="va">field</span>, <span class="va">limit</span>, <span class="va">entities</span>, <span class="va">filters</span>, <span class="va">query</span><span class="op">)</span></span></code></pre>
<ul>
<li>dataset: Any dataset with query permissions</li>
<li>field: The field within the dataset</li>
<li>limit (optional): The number of terms to retrieve (default: 10)</li>
<li>entities (optional): A list of entity tuples: [(entity_type,
entity)]</li>
<li>filters (optional): Dataset filters</li>
<li>query (optional): A query string</li>
</ul>
</div>
<div class="section level5">
<h5 id="examples-3">Examples<a class="anchor" aria-label="anchor" href="#examples-3"></a>
</h5>
<p><code>dataset_field_top_terms("quartzbio:public:/ClinVar/5.1.0-20200720/Variants-GRCh38", "clinical_significance")</code></p>
</div>
</div>
<div class="section level3">
<h3 id="dataset_field_values">dataset_field_values<a class="anchor" aria-label="anchor" href="#dataset_field_values"></a>
</h3>
<p>Retrieves a list of non-empty values for a dataset field. Returns a
list of values from the specified field. Output data type: auto
(list)</p>
<div class="section level5">
<h5 id="syntax-12">Syntax<a class="anchor" aria-label="anchor" href="#syntax-12"></a>
</h5>
<pre><code><span><span class="fu">dataset_field_values</span><span class="op">(</span><span class="va">dataset</span>, <span class="va">field</span>, <span class="va">limit</span>, <span class="va">entities</span>, <span class="va">filters</span>, <span class="va">query</span><span class="op">)</span></span></code></pre>
<ul>
<li>dataset: Any dataset with query permissions</li>
<li>field: The field within the dataset</li>
<li>limit (optional): The number of values to return (default: 10)</li>
<li>entities (optional): A list of entity tuples: [(entity_type,
entity)]</li>
<li>filters (optional): Dataset filters</li>
<li>query (optional): A query string</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="dataset_query">dataset_query<a class="anchor" aria-label="anchor" href="#dataset_query"></a>
</h3>
<p>Query any dataset with optional filters and/or entities. Returns a
list of results. Output data type: object (list)</p>
<div class="section level5">
<h5 id="syntax-13">Syntax<a class="anchor" aria-label="anchor" href="#syntax-13"></a>
</h5>
<pre><code><span><span class="fu">dataset_query</span><span class="op">(</span><span class="va">dataset</span>, <span class="va">fields</span>, <span class="va">limit</span>, <span class="va">entities</span>, <span class="va">filters</span>, <span class="va">query</span><span class="op">)</span></span></code></pre>
<ul>
<li>dataset: Any dataset with query permissions</li>
<li>fields (optional): Fields to retrieve (default: all)</li>
<li>limit (optional): The number of values to return (default: 1)</li>
<li>entities (optional): A list of entity tuples: [(entity_type,
entity)]</li>
<li>filters (optional): Dataset filters</li>
<li>query (optional): A query string</li>
</ul>
</div>
<div class="section level5">
<h5 id="examples-4">Examples<a class="anchor" aria-label="anchor" href="#examples-4"></a>
</h5>
<p><code>dataset_query("quartzbio:public:/ClinVar/5.1.0-20200720/Variants-GRCh38", fields=["clinical_significance"], query="*cancer*")</code></p>
<p><code>dataset_query("quartzbio:public:/ClinVar/5.1.0-20200720/Variants-GRCh38", entities=[["variant", "GRCH38-13-32357842-32357842-TA"]])</code></p>
</div>
</div>
<div class="section level3">
<h3 id="datetime_format">datetime_format<a class="anchor" aria-label="anchor" href="#datetime_format"></a>
</h3>
<p>Format datetime strings. By default, it returns an ISO 8601 format
date time string. To override, provide an optional input_format or
output_format to be used. Output data type: string</p>
<div class="section level5">
<h5 id="syntax-14">Syntax<a class="anchor" aria-label="anchor" href="#syntax-14"></a>
</h5>
<pre><code><span><span class="fu">datetime_format</span><span class="op">(</span><span class="va">value</span>, <span class="va">input_format</span>, <span class="va">output_format</span><span class="op">)</span></span></code></pre>
<ul>
<li>value: (str) A string containing a date/time stamp</li>
<li>input_format: (str) The input format of the date (e.g. “%d/%m/%y
%H:%M”)</li>
<li>output_format: (str) The output format of the date (ISO 8601 format
is the default: “%Y-%m-%dT%H:%M:%S”)</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="entity_ids">entity_ids<a class="anchor" aria-label="anchor" href="#entity_ids"></a>
</h3>
<p>Retrieve one or more entity IDs for a query. Output data type:
string</p>
<div class="section level5">
<h5 id="syntax-15">Syntax<a class="anchor" aria-label="anchor" href="#syntax-15"></a>
</h5>
<pre><code><span><span class="fu">entity_ids</span><span class="op">(</span><span class="va">entity_type</span>, <span class="va">entity</span><span class="op">)</span></span></code></pre>
<ul>
<li>entity_type: The entity type to retrieve</li>
<li>entity: The entity or query string</li>
</ul>
</div>
<div class="section level5">
<h5 id="error">error<a class="anchor" aria-label="anchor" href="#error"></a>
</h5>
<p>Raise a FunctionError. Output data type: error</p>
</div>
<div class="section level5">
<h5 id="syntax-16">Syntax<a class="anchor" aria-label="anchor" href="#syntax-16"></a>
</h5>
<pre><code><span><span class="fu">error</span><span class="op">(</span><span class="va">message</span><span class="op">)</span></span></code></pre>
<ul>
<li>message: An error message to raise</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="explode">explode<a class="anchor" aria-label="anchor" href="#explode"></a>
</h3>
<p>Split N values from M list fields into N records. If _id is in the
original record, each new record will have an integer appended to the
_id with the index of each exploded record. Output data type: object
(list)</p>
<div class="section level5">
<h5 id="syntax-17">Syntax<a class="anchor" aria-label="anchor" href="#syntax-17"></a>
</h5>
<pre><code><span><span class="fu">explode</span><span class="op">(</span><span class="va">record</span>, <span class="va">fields</span><span class="op">)</span></span></code></pre>
<ul>
<li>record: (object) The record to be splitted</li>
<li>fields: (list or tuple) the fields IDs</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="findall">findall<a class="anchor" aria-label="anchor" href="#findall"></a>
</h3>
<p>Returns all non-overlapping matches of pattern in string, as a list
of strings. The string is scanned left-to-right, and matches are
returned in the order found. If one or more groups are present in the
pattern, returns a list of groups. Output data type: string (list)</p>
<div class="section level5">
<h5 id="syntax-18">Syntax<a class="anchor" aria-label="anchor" href="#syntax-18"></a>
</h5>
<pre><code><span><span class="fu">findall</span><span class="op">(</span><span class="va">pattern</span>, <span class="va">string</span>, <span class="va">regex_ignorecase</span>, <span class="va">regex_dotall</span>, <span class="va">regex_multiline</span><span class="op">)</span></span></code></pre>
<ul>
<li>pattern: The regular expression pattern</li>
<li>string: The string to search</li>
<li>regex_ignorecase (default: None): With a “regex” pattern, will
perform a case insensitive matching.</li>
<li>regex_dotall (default: None): With a “regex” pattern, will make the
“.” special character match any character at all, including a newline;
without this flag, “.” will match anything except a newline.</li>
<li>regex_multiline (default: None): With a “regex” pattern, when
specified, the pattern character “^” matches at the beginning of the
string and at the beginning of each line (immediately following each
newline); and the pattern character “” matches at the end of the string
and at the end of each line (immediately preceding each newline). By
default, “^” matches only at the beginning of the string, and “” only at
the end of the string and immediately before the newline (if any) at the
end of the string.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="genomic_sequence">genomic_sequence<a class="anchor" aria-label="anchor" href="#genomic_sequence"></a>
</h3>
<p>Retrieves a specific sequence from the genome. Output data type:
string</p>
<div class="section level5">
<h5 id="syntax-19">Syntax<a class="anchor" aria-label="anchor" href="#syntax-19"></a>
</h5>
<pre><code><span><span class="fu">genomic_sequence</span><span class="op">(</span><span class="va">genomic_region</span><span class="op">)</span></span></code></pre>
<ul>
<li>genomic_region: A valid genomic region in the form:
BUILD-CHROMOSOME-START-STOP</li>
</ul>
</div>
<div class="section level5">
<h5 id="examples-5">Examples<a class="anchor" aria-label="anchor" href="#examples-5"></a>
</h5>
<p><code>genomic_sequence("GRCh37-5-36241400-36241700")</code></p>
</div>
</div>
<div class="section level3">
<h3 id="get">get<a class="anchor" aria-label="anchor" href="#get"></a>
</h3>
<p>Get the value at any depth of a nested object based on the path
described by path. If path doesn’t exist, default is returned. Output
data type: auto</p>
<div class="section level5">
<h5 id="syntax-20">Syntax<a class="anchor" aria-label="anchor" href="#syntax-20"></a>
</h5>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/get.html" class="external-link">get</a></span><span class="op">(</span><span class="va">obj</span>, <span class="va">path</span>, <span class="va">default</span><span class="op">)</span></span></code></pre>
<ul>
<li>obj: (list|dict) The object to process</li>
<li>path: (str|list) List or . delimited string of path describing
path.</li>
<li>default (keyword): Default value to return if path doesn’t exist.
Defaults to None.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="melt">melt<a class="anchor" aria-label="anchor" href="#melt"></a>
</h3>
<p>Convert a wide dataset to a long dataset by “melting” one or more
fields into “key” and “value” fields. All fields must have the same data
type. Output data type: object (list)</p>
<div class="section level5">
<h5 id="syntax-21">Syntax<a class="anchor" aria-label="anchor" href="#syntax-21"></a>
</h5>
<pre><code><span><span class="fu">melt</span><span class="op">(</span><span class="va">record</span>, <span class="va">fields</span>, <span class="va">key_field</span>, <span class="va">value_field</span>, <span class="va">melt_list_values</span><span class="op">)</span></span></code></pre>
<ul>
<li>record: (object) The record to be melted</li>
<li>fields: (list or tuple) the fields IDs</li>
<li>key_field: (str) key field (default: “key”)</li>
<li>value_field: (str) value field (default: “value”)</li>
<li>melt_list_values: (bool) (default: False)</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="normalize_aa_change">normalize_aa_change<a class="anchor" aria-label="anchor" href="#normalize_aa_change"></a>
</h3>
<p>Normalize an amino acid change (beta). Output data type: string</p>
<div class="section level5">
<h5 id="syntax-22">Syntax<a class="anchor" aria-label="anchor" href="#syntax-22"></a>
</h5>
<pre><code><span><span class="fu">normalize_aa_change</span><span class="op">(</span><span class="va">aa_change</span>, <span class="va">ref</span>, <span class="va">alt</span><span class="op">)</span></span></code></pre>
<ul>
<li>aa_change: The aa_change</li>
<li>ref: (optional) Reference allele</li>
<li>alt: (optional) Alternate allele</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="normalize_variant">normalize_variant<a class="anchor" aria-label="anchor" href="#normalize_variant"></a>
</h3>
<p>Normalize a variant ID (minimal representation and left shifting).
Output data type: string</p>
<div class="section level5">
<h5 id="syntax-23">Syntax<a class="anchor" aria-label="anchor" href="#syntax-23"></a>
</h5>
<pre><code><span><span class="fu">normalize_variant</span><span class="op">(</span><span class="va">variant</span><span class="op">)</span></span></code></pre>
<ul>
<li>variant: The variant</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="now">now<a class="anchor" aria-label="anchor" href="#now"></a>
</h3>
<p>Retrieves the current date and time. Output data type: string</p>
<div class="section level5">
<h5 id="syntax-24">Syntax<a class="anchor" aria-label="anchor" href="#syntax-24"></a>
</h5>
<pre><code><span><span class="fu">now</span><span class="op">(</span><span class="va">timezone</span>, <span class="va">template</span><span class="op">)</span></span></code></pre>
<ul>
<li>timezone (default: EST): The timezone to use for the date</li>
<li>template (default: ISO 8601): The format in which to represent the
date/time, defaults to ISO 8601 format (%Y-%m-%dT%H:%M:%S)</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="predict_variant_effects">predict_variant_effects<a class="anchor" aria-label="anchor" href="#predict_variant_effects"></a>
</h3>
<p>Predict the effects of a variant using Veppy. Output data type:
object (list). Output object properties:</p>
<ul>
<li>so_term: The Sequence Ontology term</li>
<li>impact: The effect impact</li>
<li>so_accession: The Sequence Ontology accession number</li>
<li>transcript: The affected transcript ID</li>
<li>lof: True if the mutation is predicted to cause the protein to lose
its function</li>
</ul>
<div class="section level5">
<h5 id="syntax-25">Syntax<a class="anchor" aria-label="anchor" href="#syntax-25"></a>
</h5>
<pre><code><span><span class="fu">predict_variant_effects</span><span class="op">(</span><span class="va">variant</span>, <span class="va">default_transcript</span>, <span class="va">gene_model</span><span class="op">)</span></span></code></pre>
<ul>
<li>variant: The variant</li>
<li>default_transcript (optional): If True, return effects for just the
default transcript. If a specific transcript, then limits results to
this transcript only. Otherwise returns effects for all
transcripts.</li>
<li>gene_model (optional): The desired gene model: refseq (default) or
ensembl</li>
</ul>
</div>
<div class="section level5">
<h5 id="examples-6">Examples<a class="anchor" aria-label="anchor" href="#examples-6"></a>
</h5>
<p><code>predict_variant_effects("GRCH38-7-117559590-117559593-A")</code></p>
</div>
</div>
<div class="section level3">
<h3 id="prevalence">prevalence<a class="anchor" aria-label="anchor" href="#prevalence"></a>
</h3>
<p>Calculates the frequency that a value occurs within a population.
Typically used to calculate the prevalence of variants or genes across
samples in a dataset. Returns the frequency of occurrence. Please note:
in large datasets, the result is approximate and can have an error of up
to 5%. Output data type: double</p>
<div class="section level5">
<h5 id="syntax-26">Syntax<a class="anchor" aria-label="anchor" href="#syntax-26"></a>
</h5>
<pre><code><span><span class="fu">prevalence</span><span class="op">(</span><span class="va">dataset</span>, <span class="va">entity</span>, <span class="va">sample_field</span>, <span class="va">filters</span><span class="op">)</span></span></code></pre>
<ul>
<li>dataset: Any dataset with discover permissions</li>
<li>entity: A single entity tuple: (entity_type, entity)</li>
<li>sample_field: The field containing the sample IDs</li>
<li>filters (optional): Filters to apply on the dataset</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="search">search<a class="anchor" aria-label="anchor" href="#search"></a>
</h3>
<p>Scan through string looking for the first location where the regular
expression pattern produces a match. Returns True on a match and False
if no position in the string matches the pattern. Output data type:
boolean</p>
<div class="section level5">
<h5 id="syntax-27">Syntax<a class="anchor" aria-label="anchor" href="#syntax-27"></a>
</h5>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/search.html" class="external-link">search</a></span><span class="op">(</span><span class="va">pattern</span>, <span class="va">string</span>, <span class="va">regex_ignorecase</span>, <span class="va">regex_dotall</span>, <span class="va">regex_multiline</span><span class="op">)</span></span></code></pre>
<ul>
<li>pattern: The regular expression pattern</li>
<li>string: The string to search</li>
<li>regex_ignorecase (default: None): With a “regex” pattern, will
perform a case insensitive matching.</li>
<li>regex_dotall (default: None): With a “regex” pattern, will make the
“.” special character match any character at all, including a newline;
without this flag, “.” will match anything except a newline.</li>
<li>regex_multiline (default: None): With a “regex” pattern, when
specified, the pattern character “^” matches at the beginning of the
string and at the beginning of each line (immediately following each
newline); and the pattern character “” matches at the end of the string
and at the end of each line (immediately preceding each newline). By
default, “^” matches only at the beginning of the string, and “” only at
the end of the string and immediately before the newline (if any) at the
end of the string.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="search_groups">search_groups<a class="anchor" aria-label="anchor" href="#search_groups"></a>
</h3>
<p>Scan through string looking for the first location where the regular
expression pattern produces a match. Returns a list of strings
corresponding to the groups in the pattern. Output data type: string
(list)</p>
<div class="section level5">
<h5 id="syntax-28">Syntax<a class="anchor" aria-label="anchor" href="#syntax-28"></a>
</h5>
<pre><code><span><span class="fu">search_groups</span><span class="op">(</span><span class="va">pattern</span>, <span class="va">string</span>, <span class="va">regex_ignorecase</span>, <span class="va">regex_dotall</span>, <span class="va">regex_multiline</span><span class="op">)</span></span></code></pre>
<ul>
<li>pattern: The regular expression pattern</li>
<li>string: The string to search</li>
<li>regex_ignorecase (default: None): With a “regex” pattern, will
perform a case insensitive matching.</li>
<li>regex_dotall (default: None): With a “regex” pattern, will make the
“.” special character match any character at all, including a newline;
without this flag, “.” will match anything except a newline.</li>
<li>regex_multiline (default: None): With a “regex” pattern, when
specified, the pattern character “^” matches at the beginning of the
string and at the beginning of each line (immediately following each
newline); and the pattern character “” matches at the end of the string
and at the end of each line (immediately preceding each newline). By
default, “^” matches only at the beginning of the string, and “” only at
the end of the string and immediately before the newline (if any) at the
end of the string.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="split">split<a class="anchor" aria-label="anchor" href="#split"></a>
</h3>
<p>Split text based on a delimiter and optionally strip whitespace.
Output data type: string (list)</p>
<div class="section level5">
<h5 id="syntax-29">Syntax<a class="anchor" aria-label="anchor" href="#syntax-29"></a>
</h5>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">value</span>, <span class="va">delimiter</span>, <span class="va">regex</span>, <span class="va">strip</span>, <span class="va">regex_ignorecase</span>, <span class="va">regex_dotall</span>, <span class="va">regex_multiline</span><span class="op">)</span></span></code></pre>
<ul>
<li>value: The string to split</li>
<li>delimiter (default: any whitespace): The character(s) to split
on</li>
<li>regex (default: None): A valid Python regular expression pattern to
split on.</li>
<li>strip (default: True): Strip whitespace from each resulting
value</li>
<li>regex_ignorecase (default: None): With a “regex” pattern, will
perform a case insensitive matching.</li>
<li>regex_dotall (default: None): With a “regex” pattern, will make the
“.” special character match any character at all, including a newline;
without this flag, “.” will match anything except a newline.</li>
<li>regex_multiline (default: None): With a “regex” pattern, when
specified, the pattern character “^” matches at the beginning of the
string and at the beginning of each line (immediately following each
newline); and the pattern character “” matches at the end of the string
and at the end of each line (immediately preceding each newline). By
default, “^” matches only at the beginning of the string, and “” only at
the end of the string and immediately before the newline (if any) at the
end of the string.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="sub">sub<a class="anchor" aria-label="anchor" href="#sub"></a>
</h3>
<p>Return the string obtained by replacing the leftmost non-overlapping
occurrences of pattern in string by the replacement repl. If the pattern
isn’t found, string is returned unchanged. Output data type: string</p>
<div class="section level5">
<h5 id="syntax-30">Syntax<a class="anchor" aria-label="anchor" href="#syntax-30"></a>
</h5>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="va">pattern</span>, <span class="va">repl</span>, <span class="va">string</span>, <span class="va">count</span>, <span class="va">regex_ignorecase</span>, <span class="va">regex_dotall</span>, <span class="va">regex_multiline</span><span class="op">)</span></span></code></pre>
<ul>
<li>pattern: The regular expression pattern</li>
<li>repl: The string to replace matches with</li>
<li>string: The string to search</li>
<li>count: (default: 0) The maximum number of pattern occurrences to be
replaced.If zero, all occurrences will be replaces.</li>
<li>regex_ignorecase (default: None): With a “regex” pattern, will
perform a case insensitive matching.</li>
<li>regex_dotall (default: None): With a “regex” pattern, will make the
“.” special character match any character at all, including a newline;
without this flag, “.” will match anything except a newline.</li>
<li>regex_multiline (default: None): With a “regex” pattern, when
specified, the pattern character “^” matches at the beginning of the
string and at the beginning of each line (immediately following each
newline); and the pattern character “” matches at the end of the string
and at the end of each line (immediately preceding each newline). By
default, “^” matches only at the beginning of the string, and “” only at
the end of the string and immediately before the newline (if any) at the
end of the string.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="tabulate">tabulate<a class="anchor" aria-label="anchor" href="#tabulate"></a>
</h3>
<p>Converts a list of objects into a table (i.e. a two-dimensional
array). Output data type: object (list)</p>
<div class="section level5">
<h5 id="syntax-31">Syntax<a class="anchor" aria-label="anchor" href="#syntax-31"></a>
</h5>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/tabulate.html" class="external-link">tabulate</a></span><span class="op">(</span><span class="va">objects</span>, <span class="va">fields</span>, <span class="va">header</span><span class="op">)</span></span></code></pre>
<ul>
<li>objects: The list of objects</li>
<li>fields (optional): List of fields to include (default: all)</li>
<li>header (optional): Include a header row (default: True)</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="today">today<a class="anchor" aria-label="anchor" href="#today"></a>
</h3>
<p>Returns the current date. Output data type: string</p>
<div class="section level5">
<h5 id="syntax-32">Syntax<a class="anchor" aria-label="anchor" href="#syntax-32"></a>
</h5>
<pre><code><span><span class="fu">today</span><span class="op">(</span><span class="va">timezone</span>, <span class="va">template</span><span class="op">)</span></span></code></pre>
<ul>
<li>timezone (default: EST): The timezone to use for the date</li>
<li>template (default: YYYY-MM-DD): The format in which to represent the
date</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="translate_variant">translate_variant<a class="anchor" aria-label="anchor" href="#translate_variant"></a>
</h3>
<p>Translate variant into a protein change. Output data type: object.
Output object properties:</p>
<ul>
<li>protein_length: Number of amino acids in the protein</li>
<li>cdna_change: cDNA change</li>
<li>protein_change: Protein change</li>
<li>protein_coordinates: A dictionary containing start and stop
coordinatesand the affected transcript id</li>
<li>gene: HUGO gene symbol</li>
<li>transcript: The transcript ID</li>
<li>effects: list of effects</li>
</ul>
<div class="section level5">
<h5 id="syntax-33">Syntax<a class="anchor" aria-label="anchor" href="#syntax-33"></a>
</h5>
<pre><code><span><span class="fu">translate_variant</span><span class="op">(</span><span class="va">variant</span>, <span class="va">gene_model</span>, <span class="va">transcript</span>, <span class="va">include_effects</span><span class="op">)</span></span></code></pre>
<ul>
<li>variant: The variant</li>
<li>gene_model (optional): The desired gene model: refseq (default) or
ensembl</li>
<li>transcript (optional): Limits results to this transcript only</li>
<li>include_effects (optional): Returns the effects of the variant using
Veppy</li>
</ul>
</div>
<div class="section level5">
<h5 id="examples-7">Examples<a class="anchor" aria-label="anchor" href="#examples-7"></a>
</h5>
<p><code>translate_variant("GRCH38-7-117559590-117559593-A")</code></p>
<p><code>translate_variant("GRCH38-7-117559590-117559593-A", gene_model="ensembl")</code></p>
<p><code>translate_variant("GRCH38-7-117559590-117559593-A", transcript="NM_000492.3")</code></p>
<p><code>translate_variant("GRCH38-7-117559590-117559593-A", include_effects=True)</code></p>
</div>
</div>
<div class="section level3">
<h3 id="user">user<a class="anchor" aria-label="anchor" href="#user"></a>
</h3>
<p>Returns the currently authenticated user.Output data type: object</p>
<p>Output object properties:</p>
<ul>
<li>name: The user’s full name.</li>
<li>email: The user’s email address</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Caplan, Karl Forner, QuartzBio.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Last modified: 05-Sep-2025 Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
